<h1> HDB Resale Market: Policy Analysis </h1>

---

In [1]:
import numpy as np
import pandas as pd
import itertools
import requests
import warnings

# Visualisation libraries
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import seaborn as sns
import matplotlib.pyplot as plt

# Statistic libraries
from scipy.stats import median_test
import pymannkendall as mk
import statsmodels.formula.api as smf
import pmdarima as pm
from statsmodels.tsa.statespace.sarimax import SARIMAX

# Geocoding libraries
from geopy.distance import geodesic

In [2]:
# Importing of cleaned HDB dataset into python
HDB_data_final = pd.read_csv('./Datasets/HDB_data_final_compressed.csv.gz', compression = 'gzip')

# Question A*

Yishun has received a negative reputation as “Crazy Town”, and property prices might have been impacted. Are Yishun flats the cheapest in the country?<br>

## 1. Reframing the Question for Data Analysis

The data analysis should address the following points:

<b><i>What does “cheapest” mean? </b></i><br>
Evaluate median resale price of Yishun flats as an outcome metric <br>
Evaluate price per sqm of Yishun flats to control for differences in flat size <br>
Statistically significant price difference <br>

<b><i>What is the unit of comparison? </b></i><br>
Compare Yishun to all other towns in Singapore<br>

<b><i>What is the time frame? </b></i><br>
Analyze data over the last 5 years and also historically to avoid outdated conclusions<br>

Data Analysis Question: <b><i> Do flats in Yishun have the lowest median resale price or price per sqm, that is statistically significant, compared to other towns in Singapore over the last 5 or 35 years?

## 2. Data Analysis

### 2.1. Analysis of median resale prices from 1990 to 2025

Let's first compare median resale prices and median price per sqm for towns across the entire dataset, from 1990 to 2025.

In [3]:
median_resale_town = HDB_data_final.groupby(['town'])['resale_price'].median().sort_values(ascending = False).reset_index()

In [4]:
median_resale_town['color'] = np.where(median_resale_town['town'] == 'YISHUN', 'red', 'blue')

# Bar plot of median resale price by town
fig1 = px.bar(median_resale_town,
              x = 'town',
              y = 'resale_price',
              color = 'color',
              color_discrete_map = {'red': '#d62728', 'blue': '#1f77b4'},
              width = 800, 
              height = 500)

fig1.update_layout(title = {'text' : 'HDB Median Resale Price by Town (1990 to 2025)', 'x' : 0.5, 'xanchor' : 'center'}, 
                   xaxis = {'categoryorder' : 'array', 'categoryarray': median_resale_town['town']},
                   xaxis_title_text = '', 
                   yaxis_title_text = 'Median Resale Price (S$)',
                   showlegend = False)

fig1.update_xaxes(tickangle = -90)

fig1.update_traces(hovertemplate ='Town: %{x}<br>Median Resale Price: S$%{y}<br><extra></extra>')
fig1.show()

In terms of median resale price, Yishun flats are not the lowest in the country.

In [5]:
yishun_prices = HDB_data_final.loc[HDB_data_final['town'] == 'YISHUN', 'resale_price']
towns = HDB_data_final['town'].unique()
results = []

# Perform pairwise Mood's Median Test between median resale price of Yishun and every other Town
for town in towns:
    town_prices = HDB_data_final.loc[HDB_data_final['town'] == town, 'resale_price']
    median_resale = town_prices.median()
    chi_square, p, overall_median, contingency_table = median_test(yishun_prices, town_prices)
    
    results.append({'Town' : town, 
                    'Median Resale Price (S$)' : f'{median_resale:.0f}', 
                    'p-value vs Yishun' : p,
                    'Significant': '--' if town == 'YISHUN' else ('Yes' if p < 0.05 else 'No')})
    
results_df = pd.DataFrame(results).sort_values('Median Resale Price (S$)').reset_index(drop = True)
results_df = results_df.head(10).style.map(lambda val: 'color: red' if (val >= 0.05) and (val != 1) else '', subset = ['p-value vs Yishun'])
results_df = results_df.map(lambda val: 'color: red' if val == 'No' else '', subset = ['Significant'])
results_df

,Town,Median Resale Price (S$),p-value vs Yishun,Significant
0,ANG MO KIO,230000,0.000000,Yes
1,QUEENSTOWN,235000,0.000000,Yes
2,BEDOK,250000,0.001352,Yes
3,GEYLANG,250000,0.002335,Yes
4,CLEMENTI,252000,0.340193,No
5,YISHUN,254000,1.000000,--
6,BUKIT BATOK,260000,0.000000,Yes
7,TOA PAYOH,268000,0.000000,Yes
8,JURONG EAST,270000,0.000000,Yes
9,JURONG WEST,290000,0.000000,Yes


To evaluate using statistics if median prices were significantly different, I carried out a pairwise Mood's Median Test between Yishun and every other Town. 

p-value >= 0.05 for pair-wise comparison between Yishun and Clementi, indicating that the difference in median resale price between Yishun and Clementi was not statistically significant.

In [6]:
# Create a new column in the dataset for price per sqm
HDB_data_final['price_per_sqm'] = HDB_data_final['resale_price'] / HDB_data_final['floor_area_sqm']

median_pricepersqm_town = HDB_data_final.groupby(['town'])['price_per_sqm'].median().sort_values(ascending = False).reset_index()

In [8]:
median_pricepersqm_town['color'] = np.where(median_pricepersqm_town['town'] == 'YISHUN', 'red', 'blue')

# Bar plot of median resale price by town
fig2 = px.bar(median_pricepersqm_town,
              x = 'town',
              y = 'price_per_sqm',
              color = 'color',
              color_discrete_map = {'red': '#d62728', 'blue': '#1f77b4'},
              width = 800, 
              height = 500)

fig2.update_layout(title = {'text' : 'HDB Median Resale Price per SQM by Town (1990 to 2025)', 'x' : 0.5, 'xanchor' : 'center'}, 
                   xaxis = {'categoryorder' : 'array', 'categoryarray': median_pricepersqm_town['town']},
                   xaxis_title_text = '', 
                   yaxis_title_text = 'Median Price per SQM (S$/m²)',
                   showlegend = False)

fig2.update_xaxes(tickangle = -90)

fig2.update_traces(hovertemplate ='Town: %{x}<br>Median Price per SQM: S$%{y:.0f} per m²<br><extra></extra>')
fig2.show()

This is likely because Yishun has many larger resale flats such as older 4- and 5-room flats. Larger flats have a higher resale price but lower price per sqm. 

In [9]:
yishun_prices = HDB_data_final.loc[HDB_data_final['town'] == 'YISHUN', 'price_per_sqm']
towns = HDB_data_final['town'].unique()
results = []

# Perform pairwise Mood's Median Test between median resale price per sqm of Yishun and every other Town
for town in towns:
    town_prices = HDB_data_final.loc[HDB_data_final['town'] == town, 'price_per_sqm']
    median_resale = town_prices.median()
    chi_square, p, overall_median, contingency_table = median_test(yishun_prices, town_prices)
    
    results.append({'Town' : town, 
                    'Median Price per SQM (S$/m<sup>2</sup>)' : f'{median_resale:.0f}', 
                    'p-value vs Yishun' : p,
                    'Significant': '--' if town == 'YISHUN' else ('Yes' if p < 0.05 else 'No')})

results_df = pd.DataFrame(results).sort_values('Median Price per SQM (S$/m<sup>2</sup>)').reset_index(drop = True)
results_df = results_df.head(10).style.map(lambda val: 'color: red' if (val >= 0.05) and (val != 1) else '', subset = ['p-value vs Yishun'])
results_df = results_df.map(lambda val: 'color: red' if val == 'No' else '', subset = ['Significant'])
results_df

,Town,Median Price per SQM (S$/m2),p-value vs Yishun,Significant
0,YISHUN,2603,1.000000,--
1,BUKIT BATOK,2657,0.000000,Yes
2,JURONG EAST,2672,0.000000,Yes
3,WOODLANDS,2677,0.000000,Yes
4,JURONG WEST,2708,0.000000,Yes
5,ANG MO KIO,2733,0.000000,Yes
6,BEDOK,2769,0.000000,Yes
7,HOUGANG,2908,0.000000,Yes
8,CLEMENTI,2910,0.000000,Yes
9,GEYLANG,2932,0.000000,Yes


p-value < 0.05 for all pair-wise comparisons, indicating that the difference in median resale price per sqm between Yishun and every other town were statistically significant. <br>

In [10]:
# Create a new column in the dataset for price per sqm
top5cheapesttowns = ['YISHUN', 'CLEMENTI', 'GEYLANG', 'BEDOK', 'ANG MO KIO', 'QUEENSTOWN']

median_floorarea_town = HDB_data_final[(HDB_data_final['town'].isin(top5cheapesttowns))]
median_floorarea_town = median_floorarea_town.groupby(['transaction_year', 'town'])['floor_area_sqm'].median().reset_index()

In [12]:
# Line plot of median floor area in the top 5 towns ranked by the number of HDB flats sold from 1990 to 2025
fig3 = px.line(median_floorarea_town,
               x = 'transaction_year',
               y = 'floor_area_sqm',
               color = 'town',
               width = 800, 
               height = 500,
               markers = True)

fig3.update_layout(title = {'text' : 'Median Floor Area of Resale Flats in the Top 6 Cheapest Towns', 'x' : 0.45, 'xanchor' : 'center'}, 
                   xaxis_title_text = 'Transaction Year', 
                   yaxis_title_text = 'Median Floor Area (m<sup>2</sup>)',
                   showlegend = True)

fig3.update_traces(hovertemplate = 'Transaction Year: %{x}<br>Median Floor Area: %{y} m<sup>2</sup><br>Town: %{fullData.name}<extra></extra>',
                   marker = dict(size = 8))

fig3.update_layout(legend_title_text = 'Town')

fig3.show()

Amongst the towns with the lowest median resale price, Yishun flats have the highest median floor area. This explains why flats in Yishun have the lowest price per sqm in the country.

### 2.1. Analysis of median resale prices from 2021 to 2025

Next, I will analyze more recent resale price data from 2021 to 2025.

In [53]:
# Filter to look at data from 2021 to 2025
HDB_data_recent = HDB_data_final.loc[(HDB_data_final['transaction_year'] > 2020) & (HDB_data_final['transaction_year'] <= 2025)]

median_resale_town = HDB_data_recent.groupby(['town'])['resale_price'].median().sort_values(ascending = False).reset_index()

In [54]:
median_resale_town['color'] = np.where(median_resale_town['town'] == 'YISHUN', 'red', 'blue')

# Bar plot of median resale price by town
fig4 = px.bar(median_resale_town,
              x = 'town',
              y = 'resale_price',
              color = 'color',
              color_discrete_map = {'red': '#d62728', 'blue': '#1f77b4'},
              width = 800, 
              height = 500)

fig4.update_layout(title = {'text' : 'HDB Median Resale Price by Town (2021 to 2025)', 'x' : 0.5, 'xanchor' : 'center'}, 
                   xaxis = {'categoryorder' : 'array', 'categoryarray': median_resale_town['town']},
                   xaxis_title_text = '', 
                   yaxis_title_text = 'Median Resale Price (S$)',
                   showlegend = False)

fig4.update_xaxes(tickangle = -90)

fig4.update_traces(hovertemplate ='Town: %{x}<br>Median Resale Price: S$%{y}<br><extra></extra>')
fig4.show()

In [55]:
yishun_prices = HDB_data_recent.loc[HDB_data_recent['town'] == 'YISHUN', 'resale_price']
towns = HDB_data_recent['town'].unique()
results = []

# Perform pairwise Mood's Median Test between median resale price of Yishun and every other Town
for town in towns:
    town_prices = HDB_data_recent.loc[HDB_data_recent['town'] == town, 'resale_price']
    median_resale = town_prices.median()
    chi_square, p, overall_median, contingency_table = median_test(yishun_prices, town_prices)
    
    results.append({'Town' : town, 
                    'Median Resale Price (S$)' : f'{median_resale:.0f}', 
                    'p-value vs Yishun' : p,
                    'Significant': '--' if town == 'YISHUN' else ('Yes' if p < 0.05 else 'No')})

results_df = pd.DataFrame(results).sort_values('Median Resale Price (S$)').reset_index(drop = True)
results_df = results_df.head(10).style.map(lambda val: 'color: red' if (val >= 0.05) and (val != 1) else '', subset = ['p-value vs Yishun'])
results_df = results_df.map(lambda val: 'color: red' if val == 'No' else '', subset = ['Significant'])
results_df

,Town,Median Resale Price (S$),p-value vs Yishun,Significant
0,ANG MO KIO,443000,0.000000,Yes
1,BEDOK,470000,0.006864,Yes
2,JURONG EAST,473000,0.141415,No
3,YISHUN,480000,1.000000,--
4,MARINE PARADE,506000,0.000018,Yes
5,GEYLANG,508000,0.000045,Yes
6,JURONG WEST,510000,0.000000,Yes
7,WOODLANDS,525000,0.000000,Yes
8,CHOA CHU KANG,528844,0.000000,Yes
9,BUKIT PANJANG,530000,0.000000,Yes


p-value >= 0.05 for pair-wise comparison between Yishun and Jurong East, indicating that the difference in median resale price between Yishun and Jurong East was not statistically significant.

In [56]:
median_pricepersqm_town = HDB_data_recent.groupby(['town'])['price_per_sqm'].median().sort_values(ascending = False).reset_index()

In [ ]:
median_pricepersqm_town['color'] = np.where(median_pricepersqm_town['town'] == 'YISHUN', 'red', 'blue')

# Bar plot of median resale price by town
fig5 = px.bar(median_pricepersqm_town,
              x = 'town',
              y = 'price_per_sqm',
              color = 'color',
              color_discrete_map = {'red' : '#d62728', 'blue' : '#1f77b4'},
              width = 800, 
              height = 500)

fig5.update_layout(title = {'text' : 'HDB Median Resale Price per SQM by Town (2021 to 2025)', 'x' : 0.5, 'xanchor' : 'center'}, 
                   xaxis = {'categoryorder' : 'array', 'categoryarray' : median_pricepersqm_town['town']},
                   xaxis_title_text = '', 
                   yaxis_title_text = 'Median Price per SQM (S$/m²)',
                   showlegend = False)

fig5.update_xaxes(tickangle = -90)

fig5.update_traces(hovertemplate ='Town: %{x}<br>Median Price per SQM: S$%{y:.0f} per m²<br><extra></extra>')
fig5.show()

The median price per sqm of Yishun flats have risen significantly in recent years and are no longer the lowest in the country.

In [58]:
yishun_prices = HDB_data_recent.loc[HDB_data_recent['town'] == 'YISHUN', 'price_per_sqm']
towns = HDB_data_recent['town'].unique()
results = []

# Perform pairwise Mood's Median Test between median price_per_sqm of Yishun and every other Town
for town in towns:
    town_prices = HDB_data_recent.loc[HDB_data_recent['town'] == town, 'price_per_sqm']
    median_resale = town_prices.median()
    chi_square, p, overall_median, contingency_table = median_test(yishun_prices, town_prices)
    
    results.append({'Town' : town, 
                    'Median Price per SQM (S$/m<sup>2</sup>)' : f'{median_resale:.0f}', 
                    'p-value vs Yishun' : p,
                    'Significant': '--' if town == 'YISHUN' else ('Yes' if p < 0.05 else 'No')})
    
results_df = pd.DataFrame(results).sort_values('Median Price per SQM (S$/m<sup>2</sup>)').reset_index(drop = True)
results_df = results_df.head(10).style.map(lambda val: 'color: red' if (val >= 0.05) and (val != 1) else '', subset = ['p-value vs Yishun'])
results_df = results_df.map(lambda val: 'color: red' if val == 'No' else '', subset = ['Significant'])
results_df

,Town,Median Price per SQM (S$/m2),p-value vs Yishun,Significant
0,JURONG WEST,5041,0.000000,Yes
1,CHOA CHU KANG,5075,0.000000,Yes
2,WOODLANDS,5123,0.000000,Yes
3,JURONG EAST,5248,0.000000,Yes
4,PASIR RIS,5274,0.000000,Yes
5,BUKIT PANJANG,5328,0.000000,Yes
6,YISHUN,5435,1.000000,--
7,BEDOK,5522,0.000068,Yes
8,ANG MO KIO,5662,0.000000,Yes
9,HOUGANG,5677,0.000000,Yes


p-value < 0.05 for all pair-wise comparisons, indicating that the difference in median resale price per sqm between Yishun and every other town were statistically significant. <Br>

## 3. Conclusion

Based on my data analysis, comparing median resale prices of flats in Yishun to flats in all other towns in Singapore, Yishun flats are not the cheapest regardless of time frame being compared.

However, if we compare median resale prices per sqm, which takes into account the flat size, Yishun flats are the cheapest in the timeframe of 1990 to 2025. This is because amongst the towns with the lowest median resale price, Yishun flats have the highest median floor area. Larger flats are higher in price but lower in price per sqm.

For more recent years, from 2021 to 2025, the median resale price per sqm of Yishun flats is not the lowest in the country anymore. <br> <br>

# Question B*

Some members of public have been saying that flat sizes have gotten smaller over the years. Is there any truth in this statement?

## 1. Reframing the Question for Data Analysis

The data analysis should address the following points:

<b><i>How do we determine if flat sizes have gotten smaller? </b></i><br>
Evaluate median floor area as an outcome metric <br>
Statistically significant downward trend <br>

<b><i>What is the unit of comparison? </b></i><br>
Compare between different regions in Singapore and also on a national level<br>
Compare between different flat types<br>

<b><i>What is the time frame? </b></i><br>
Analyze data from 1990 to 2025 on a yearly granularity <br>

<b><i>Are there any limitations in the data? </b></i><br>
The dataset only includes HDB flats which were resold, not new flats purchased from HDB or flats that were leased out. <br>

Data Analysis Question: <b><i> Has there been a statistically significant decrease in median floor area of resale flats on a national level, for different regions, HDB towns and flat types, from 1990 to 2025?

## 2. Data Analysis

### 2.1. Analysis of median floor area over time on a national level

I will first analyse the trend in median floor area on a national level from 1990 to 2025.

In [59]:
median_floorarea_year = HDB_data_final.groupby(['transaction_year'])['floor_area_sqm'].median().reset_index()

In [ ]:
# Line plot of median floor area from 1990 to 2025
fig6 = px.line(median_floorarea_year,
               x = 'transaction_year',
               y = 'floor_area_sqm',
               width = 800, 
               height = 500,
               markers = True)

fig6.update_layout(title = {'text' : 'Median Floor Area of Resale Flats over Time', 'x' : 0.5, 'xanchor' : 'center'}, 
                   xaxis_title_text = 'Transaction Year', 
                   yaxis_title_text = 'Median Floor Area (m<sup>2</sup>)',
                   showlegend = False)

fig6.update_traces(hovertemplate ='Transaction Year: %{x}<br>Median Floor Area: %{y} m<sup>2</sup><br><extra></extra>',
                   marker = dict(color =' red', size = 8))
fig6.show()

From the line plot, a clear increasing trend in median floor area of resale flats can be observed from 1990 to 1999. <br>

Between 2000 to 2007, huge fluctuations were observed followed by a decrease in median floor area of resale flats from 2008 to 2025.

To evaluate using statistics if an increasing or decreasing trend in median floor area was statistically significant, I will carry out a Mann-Kendall test on each line plot. <br>

The Mann-Kendall test evaluates monotonic trends in a time series. If p-value < 0.05, it means that a statistically significant decreasing or increasing trend in median floor area was observed.

In [ ]:
# Define the 3 time frames
time_frames = {'1990-1999' : (1990, 1999),
               '2000-2007' : (2000, 2007),
               '2008-2025' : (2008, 2025)}
results = []

# Apply Mann-Kendall test for each time frame
for label, (start, end) in time_frames.items():
    subset = median_floorarea_year[(median_floorarea_year['transaction_year'] >= start) &
                                   (median_floorarea_year['transaction_year'] <= end)]
    
    result = mk.original_test(subset['floor_area_sqm'])
    results.append({'Time Frame': label,
                    'Trend': result.trend,
                    'p-value': result.p,
                    'Significant': 'Yes' if result.p < 0.05 else 'No'})

results_df = pd.DataFrame(results)
results_df = results_df.style.map(lambda val: 'color: red' if val == 'decreasing' else '', subset = ['Trend'])
results_df

,Time Frame,Trend,p-value,Significant
0,1990-1999,increasing,0.000347,Yes
1,2000-2007,no trend,0.706197,No
2,2008-2025,decreasing,0.000242,Yes


To better evaluate the trends in median floor area of resale flats, the data was further segmented into 3 time frames: 1990 to 1999, 2000 to 2007, and 2008 to 2025. 

The median floor area of resale flats for this 3 time frames will be separately evaluated using the Mann-Kendall test to test for statistically significant trends.

From 2008 to 2025, p-value < 0.05, indicating that the decreasing trend in median floor area of resale flats in that time frame  was statistically significant.

### 2.2. Analysis of median floor area over time for different regions and towns

In [62]:
town_to_region = {'ANG MO KIO' : 'Central',
                  'BISHAN' : 'Central',
                  'BUKIT MERAH' : 'Central',
                  'BUKIT TIMAH' : 'Central',
                  'CENTRAL AREA' : 'Central',
                  'GEYLANG' : 'Central',
                  'KALLANG/WHAMPOA' : 'Central',
                  'MARINE PARADE' : 'Central',
                  'QUEENSTOWN' : 'Central',
                  'TOA PAYOH' : 'Central',
                  'BEDOK' : 'East',
                  'PASIR RIS' : 'East',
                  'TAMPINES' : 'East',
                  'PUNGGOL' : 'North-East',
                  'SERANGOON' : 'North-East',
                  'SENGKANG' : 'North-East',
                  'HOUGANG' : 'North-East',
                  'WOODLANDS' : 'North',
                  'SEMBAWANG' : 'North',
                  'YISHUN' : 'North',
                  'BUKIT BATOK' : 'West',
                  'BUKIT PANJANG' : 'West',
                  'CHOA CHU KANG' : 'West',
                  'CLEMENTI' : 'West',
                  'JURONG EAST' : 'West',
                  'JURONG WEST' : 'West',
                  'WEST COAST' : 'West'}

The different HDB towns were grouped into 5 different regions in Singapore to facilitate analysis.

In [63]:
HDB_data_final['region'] = HDB_data_final['town'].map(town_to_region)

median_floorarea_region_year = HDB_data_final.groupby(['transaction_year', 'region'])['floor_area_sqm'].median().reset_index()

In [ ]:
# Line plot of median floor area for different regions in Singapore from 1990 to 2025
fig7 = px.line(median_floorarea_region_year,
               x = 'transaction_year',
               y = 'floor_area_sqm',
               color = 'region',
               width = 800, 
               height = 500,
               markers = True)

fig7.update_layout(title = {'text' : 'Median Floor Area of Resale Flats for Different Regions', 'x' : 0.465, 'xanchor' : 'center'}, 
                   xaxis_title_text = 'Transaction Year', 
                   yaxis_title_text = 'Median Floor Area (m<sup>2</sup>)',
                   showlegend = True)

fig7.update_traces(hovertemplate ='Transaction Year: %{x}<br>Median Floor Area: %{y} m<sup>2</sup><br>Region: %{fullData.name}<extra></extra>',                
                   marker = dict(size = 8))

fig7.update_layout(legend_title_text = 'Region')

fig7.show()

A clear increasing trend for median floor area of resale flats was observed for the Central region. 

For the East, West, and North regions, median floor area of resale flats increased from 1990 to 1999 but remained almost the same from 1999 to 2025.

For the North-East region, median floor area of resale flats increased quite rapidly from 1990 to 2008 but also decreased quite drastically from 2008 to 2025.

Overall, median floor areas of resale flats in different regions, with the exception of North-East Region, have increased significantly over the years from 68, 84, 82, and 90 m<sup>2</sup> in 1990 to 87, 93, 100, and 103 m<sup>2</sup> in 2025 for the Central, North, North-East, West, and East regions, respectively.

In [65]:
regions = HDB_data_final['region'].unique()
results = []

# Perform Mann-Kendall test on all line plots of flat types over time
for region in regions:
    region_data = median_floorarea_region_year[median_floorarea_region_year['region'] == region].sort_values('transaction_year')
    result = mk.original_test(region_data['floor_area_sqm'])
    
    results.append({'Region' : region,
                    'Trend' : result.trend,
                    'p-value' : result.p,
                    'Significant' : 'Yes' if result.p < 0.05 else 'No'})

results_df = pd.DataFrame(results).sort_values('Region').reset_index(drop = True)
results_df = results_df.style.map(lambda val: 'color: red' if val == 'decreasing' else '', subset = ['Trend'])
results_df

,Region,Trend,p-value,Significant
0,Central,increasing,0.000000,Yes
1,East,no trend,0.103006,No
2,North,no trend,0.347119,No
3,North-East,no trend,0.719301,No
4,West,increasing,0.009365,Yes


Out of the 5 line plots for different regions in Singapore, p-value < 0.05 for the Central and West regions. This indicates that the increasing trend in median floor area of resale flats in the Central and West regions from 1990 to 2025 were statistically significant.

No statistically significant decrease in median floor area of resale flats was observed for any regions.

In [66]:
top_towns = HDB_data_final['town'].value_counts().nlargest(5).index.tolist()

median_floorarea_town_year = HDB_data_final.groupby(['transaction_year', 'town'])['floor_area_sqm'].median().reset_index()
median_floorarea_toptowns_year = median_floorarea_town_year[median_floorarea_town_year['town'].isin(top_towns)]

The top 5 HDB towns with the highest number of HDB flats sold across the years were selected to be plotted. 

In [ ]:
# Line plot of median floor area in the top 5 towns ranked by the number of HDB flats sold from 1990 to 2025
fig8 = px.line(median_floorarea_toptowns_year,
               x = 'transaction_year',
               y = 'floor_area_sqm',
               color = 'town',
               width = 800, 
               height = 500,
               markers = True)

fig8.update_layout(title = {'text' : 'Median Floor Area of Resale Flats in the Top 5 Towns', 'x' : 0.45, 'xanchor' : 'center'}, 
                   xaxis_title_text = 'Transaction Year', 
                   yaxis_title_text = 'Median Floor Area (m<sup>2</sup>)',
                   showlegend = True)

fig8.update_traces(hovertemplate = 'Transaction Year: %{x}<br>Median Floor Area: %{y} m<sup>2</sup><br>Town: %{fullData.name}<extra></extra>',
                   marker = dict(size = 8))

fig8.update_layout(legend_title_text = 'Town')

fig8.show()

The median floor area of resale flats for Tampines remained relatively constant over the years. 

For Yishun, median floor area of resale flats saw a huge spike in 1993 but dropped back down to the same level in the years that followed.

Both Jurong West and Woodlands saw a huge spike in median floor area of resale flats from 1990 to 2000 but remained relatively unchanged after.

A more gradual increase in median floor are of resale flats was observed for Bedok.

In [68]:
results = []

# Perform Mann-Kendall test on all line plots of all 5 towns over time
for town in top_towns:
    town_data = median_floorarea_toptowns_year[median_floorarea_toptowns_year['town'] == town].sort_values('transaction_year')
    result = mk.original_test(town_data['floor_area_sqm'])
    
    results.append({'Town' : town,
                    'Trend' : result.trend,
                    'p-value' : result.p,
                    'Significant' : 'Yes' if result.p < 0.05 else 'No'})

results_df = pd.DataFrame(results).sort_values('Town').reset_index(drop = True)
results_df = results_df.style.map(lambda val: 'color: red' if val == 'decreasing' else '', subset = ['Trend'])
results_df

,Town,Trend,p-value,Significant
0,BEDOK,increasing,0.003347,Yes
1,JURONG WEST,increasing,0.002139,Yes
2,TAMPINES,no trend,0.559426,No
3,WOODLANDS,increasing,0.005399,Yes
4,YISHUN,no trend,0.101384,No


Out of the 5 line plots for the 5 biggest towns Singapore, p-value < 0.05 for Bedok, Jurong West, and Woodlands. This indicates that the increasing trend in median floor area of resale flats in Bedok, Jurong West, and Woodlands from 1990 to 2025 were statistically significant.

No statistically significant decrease in median floor area of resale flats was observed for any of these towns.

### 2.3. Analysis of median floor area over time for different flat types

In [ ]:
median_floorarea_flattype_year = HDB_data_final.groupby(['transaction_year', 'flat_type'])['floor_area_sqm'].median().reset_index()

In [ ]:
# Line plot of median floor area for different flat types from 1990 to 2025
fig9 = px.line(median_floorarea_flattype_year,
               x = 'transaction_year',
               y = 'floor_area_sqm',
               color = 'flat_type',
               width = 850, 
               height = 500,
               markers = True)

fig9.update_layout(title = {'text' : 'Median Floor Area of Resale Flats for Different Flat Types', 'x' : 0.43, 'xanchor' : 'center'}, 
                   xaxis_title_text = 'Transaction Year', 
                   yaxis_title_text = 'Median Floor Area (m<sup>2</sup>)',
                   showlegend = True)

fig9.update_traces(hovertemplate ='Transaction Year: %{x}<br>Median Floor Area: %{y} m<sup>2</sup><br>Flat Type: %{fullData.name}<extra></extra>',             
                   marker = dict(size = 8))

fig9.update_layout(legend_title_text = 'Flat Type')

fig9.show()

The median floor area of resale flats for 1 room, 2 room, 3 room, and multi-generation flat types have remained relatively unchanged over the years. 

The median floor area of resale flats of 5 room and executive flats experienced a gradual decline from 1990 to 2025, from 121 m<sup>2</sup> to 116 m<sup>2</sup> and from 149 m<sup>2</sup> to 146 m<sup>2</sup>, respectively.

The median floor area of resale flats of 4 room flats increased from 1990 to 1999, followed by a gradual decline from 1999 to 2025, from 103 m<sup>2</sup> to 93 m<sup>2</sup>.

In [ ]:
flat_types = HDB_data_final['flat_type'].unique()
results = []

# Perform Mann-Kendall test on all line plots of flat types over time
for flat_type in flat_types:
    flat_type_data = median_floorarea_flattype_year[median_floorarea_flattype_year['flat_type'] == flat_type].sort_values('transaction_year')
    result = mk.original_test(flat_type_data['floor_area_sqm'])
    
    results.append({'Flat Type' : flat_type,
                    'Trend' : result.trend,
                    'p-value' : result.p,
                    'Significant' : 'Yes' if result.p < 0.05 else 'No'})

results_df = pd.DataFrame(results).sort_values('Flat Type').reset_index(drop = True)
results_df = results_df.style.map(lambda val: 'color: red' if val == 'decreasing' else '', subset = ['Trend'])
results_df

,Flat Type,Trend,p-value,Significant
0,1 ROOM,no trend,1.000000,No
1,2 ROOM,no trend,0.531417,No
2,3 ROOM,no trend,0.908863,No
3,4 ROOM,decreasing,0.007847,Yes
4,5 ROOM,decreasing,0.000000,Yes
5,EXECUTIVE,decreasing,0.000734,Yes
6,MULTI GENERATION,no trend,0.848366,No


Out of the 7 line plots for different flat types, p-value < 0.05 for 4 room, 5 room, and executive flats. This indicates that the decreasing trend in median floor area for 4 room, 5 room, and executive resale flats from 1990 to 2025 were statistically significant. <br>

In [ ]:
median_floorarea_no5RM_year = HDB_data_final[HDB_data_final['flat_type'] != '5 ROOM']
median_floorarea_no5RM_year = median_floorarea_no5RM_year.groupby(['transaction_year'])['floor_area_sqm'].median().reset_index()
median_floorarea_no5RM_year['Excludes'] = '5 ROOM'

median_floorarea_no4RM_year = HDB_data_final[HDB_data_final['flat_type'] != '4 ROOM']
median_floorarea_no4RM_year = median_floorarea_no4RM_year.groupby(['transaction_year'])['floor_area_sqm'].median().reset_index()
median_floorarea_no4RM_year['Excludes'] = '4 ROOM'

median_floorarea_noExec_year = HDB_data_final[HDB_data_final['flat_type'] != 'EXECUTIVE']
median_floorarea_noExec_year = median_floorarea_noExec_year.groupby(['transaction_year'])['floor_area_sqm'].median().reset_index()
median_floorarea_noExec_year['Excludes'] = 'EXECUTIVE'

median_floorarea_noMG_year = HDB_data_final[HDB_data_final['flat_type'] != 'MULTI GENERATION']
median_floorarea_noMG_year = median_floorarea_noMG_year.groupby(['transaction_year'])['floor_area_sqm'].median().reset_index()
median_floorarea_noMG_year['Excludes'] = 'MULTI GENERATION'

median_floorarea_no1RM_year = HDB_data_final[HDB_data_final['flat_type'] != '1 ROOM']
median_floorarea_no1RM_year = median_floorarea_no1RM_year.groupby(['transaction_year'])['floor_area_sqm'].median().reset_index()
median_floorarea_no1RM_year['Excludes'] = '1 ROOM'

median_floorarea_no2RM_year = HDB_data_final[HDB_data_final['flat_type'] != '2 ROOM']
median_floorarea_no2RM_year = median_floorarea_no2RM_year.groupby(['transaction_year'])['floor_area_sqm'].median().reset_index()
median_floorarea_no2RM_year['Excludes'] = '2 ROOM'

median_floorarea_no3RM_year = HDB_data_final[HDB_data_final['flat_type'] != '3 ROOM']
median_floorarea_no3RM_year = median_floorarea_no3RM_year.groupby(['transaction_year'])['floor_area_sqm'].median().reset_index()
median_floorarea_no3RM_year['Excludes'] = '3 ROOM'

median_floorarea_exclusions_year = pd.concat([median_floorarea_no1RM_year, median_floorarea_no2RM_year, median_floorarea_no3RM_year, median_floorarea_no4RM_year, 
                                              median_floorarea_no5RM_year, median_floorarea_noExec_year, median_floorarea_noMG_year], axis = 0)

In [ ]:
# Line plot of median floor area from 1990 to 2025, excluding certain flat types
fig10 = px.line(median_floorarea_exclusions_year,
                x = 'transaction_year',
                y = 'floor_area_sqm',
                color = 'Excludes',
                width = 850, 
                height = 500,
                markers = True)

fig10.update_layout(title = {'text' : 'Median Floor Area of Resale Flats over Time (Excluding Certain Flat Types)', 'x' : 0.43, 'xanchor' : 'center'}, 
                    xaxis_title_text = 'Transaction Year', 
                    yaxis_title_text = 'Median Floor Area (m<sup>2</sup>)',
                    showlegend = True)

fig10.update_traces(hovertemplate ='Transaction Year: %{x}<br>Median Floor Area: %{y} m<sup>2</sup><br>Excludes: %{fullData.name}<extra></extra>',             
                    marker = dict(size = 8))

fig10.show()

Without 1 room, 2 room, 3 room, or multi generation flats, median floor area would have still trended downwards from 2008 to 2025.

Without 5 room flats, median floor area would have trended upwards instead from 2008 to 2025.

Without 4 room or executive flats, median floor area would have remained relatively constant.

In [ ]:
median_floorarea_exclusions_year = median_floorarea_exclusions_year[(median_floorarea_exclusions_year['transaction_year'] >= 2008) &
                                                                    (median_floorarea_exclusions_year['transaction_year'] <= 2025)]

exclusions = median_floorarea_exclusions_year['Excludes'].unique()
results = []

# Perform Mann-Kendall test for all line plots which excludes certain flat types for the time frame 2008 to 2025
for exclusion in exclusions:
    exclusion_data = median_floorarea_exclusions_year[median_floorarea_exclusions_year['Excludes'] == exclusion].sort_values('transaction_year')
    result = mk.original_test(exclusion_data['floor_area_sqm'])
    
    results.append({'Exclusion' : exclusion,
                    'Trend (2008 to 2025)' : result.trend,
                    'p-value' : result.p,
                    'Significant' : 'Yes' if result.p < 0.05 else 'No'})

results_df = pd.DataFrame(results).sort_values('Exclusion').reset_index(drop = True)
results_df = results_df.style.map(lambda val: 'color: red' if val != 'decreasing' else '', subset = ['Trend (2008 to 2025)'])
results_df

,Exclusion,Trend (2008 to 2025),p-value,Significant
0,1 ROOM,decreasing,0.000242,Yes
1,2 ROOM,decreasing,0.000082,Yes
2,3 ROOM,decreasing,0.000006,Yes
3,4 ROOM,no trend,0.874940,No
4,5 ROOM,increasing,0.000202,Yes
5,EXECUTIVE,no trend,0.431320,No
6,MULTI GENERATION,decreasing,0.000242,Yes


## 3. Conclusion

Based on my data analysis, across Singapore, the median floor area of resale flats experienced a statistically significant increase from 1990 to 1999. Between 2000 to 2007, no significant trend was observed. 

However, from 2008 to 2025, a statistically significant decrease in median floor area of resale flats was observed. 

Comparing between different flat types, a statistically significant decrease in median floor area of resale flats was observed for 4 room, 5 room, and executive resale flats from 1990 to 2025.
The median floor area decreased from 103 m<sup>2</sup> to 93 m<sup>2</sup>, 121 m<sup>2</sup> to 116 m<sup>2</sup> and 149 m<sup>2</sup> to 146 m<sup>2</sup> for 4 room, 5 room, and executive resale flats, respectively, from 1990 to 2025. 

Without 5 room flats, median floor area of resale flats would have trended upwards instead from 2008 to 2025.

Without 4 room or executive flats, median floor area of resale flats would have remained relatively constant from 2008 to 2025.

Therefore, there is some truth in the statement that flat sizes have decreased over the years, specifically from 2008 to 2025, and largely due to shrinking 4-room, 5-room and executive resale flats. <br> <br>


# Question C**

The Downtown Line Stage 2 connects the Bukit Panjang heartland to the city. Have prices increased for resale flats in the towns served by this Line? You might want to use a difference-in-differences model for this task. <br>

## 1. Reframing the Question for Data Analysis

The data analysis should address the following points:

<b><i>How do we determine if prices have increased? </b></i><br>
Evaluate % increase in price per sqm as an outcome metric <br>
Statistically significant price difference using the DiD regression model <br>

<b><i>What is the unit of comparison? </b></i><br>
Use flats near to the DTL Stage 2 stations as the treatment group and flats that are not as control group for the model <br>

<b><i>What is the time frame? </b></i><br>
1st model: Analyze data from 2011 to 2020, 5 years before and after the start of DTL Stage 2 on 2015 <br>
2nd model: Analyze data from 2013 to 2018, 3 years before and after the start of DTL Stage 2 on 2015 <br>

Data Analysis Question: <b><i> Has the price per sqm of flats nearest to DTL Stage 2 stations experienced a statistically significant increase relative to flats that are further away, 3 or 5 years before and after 2015 when the line started?

## 2. Data Pre-processing for DiD regression model

The geospatial coordinates of the 10 DTL2 MRT stations will be extracted via Onemap's geocoding API.

In [ ]:
# Access token received from onemap.gov.sg after registration (EXPIRED)
access_token = 'Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyX2lkIjo4Mzg3LCJmb3JldmVyIjpmYWxzZSwiaXNzIjoiT25lTWFwIiwiaWF0IjoxNzU3NjM2NDUzLCJuYmYiOjE3NTc2MzY0NTMsImV4cCI6MTc1Nzg5NTY1MywianRpIjoiMWZlMDRkZjEtYTU0MS00ZWMzLWIwNjYtYTUyZGU1NWNjMjc4In0.DK-m7EuGii9xjXxec_BOxVrSEZfE7NA7yvKVXn3UsThYAhMEHjYu5AEvsty26tx2CFR458FoxEEArHLW4z9aCRUe5-6n_LkUkdYSE87VjM1RiIlO-L6ON0S7LIQflsbRTxb8JnXLBpHPfzd6qpjn2qi6ZxKZ2_AXtqX50N6OE2c9j6YNeo055qWWmG1C6XVpsfafgcRLmJX0loschwK8K_HLEeIqtyejjIqC_gMv-7muJZqtWz7ji1nhx8YY2iJjgrb4-5a5S-2gD_wrBVDKATJ7ZDZ1zby0lNiUFy2EzttIua0XL_uWAMyIxgdy6oOwpkEklDqzLiDR9LA8j1tr1g'

In [76]:
MRT_ID = ['DT1', 'DT2', 'DT3', 'DT4', 'DT5', 'DT6', 'DT7', 'DT8', 'DT9', 'CC19']

In [77]:
MRT_latitude = []
MRT_longitude = []
MRT_Name = []

# Loop through MRT_ID list to search for the station's geospatial coordinates
for MRT_ID_row in MRT_ID:
    url = f'https://www.onemap.gov.sg/api/common/elastic/search?searchVal={MRT_ID_row}&returnGeom=Y&getAddrDetails=Y'
    response = requests.get(url, headers = {'Authorization': access_token}).json()['results']

    # Only append values to list if response does not return an empty result
    if response:
        MRT_Name.append(response[0]['BUILDING'])
        MRT_latitude.append(response[0]['LATITUDE'])
        MRT_longitude.append(response[0]['LONGITUDE'])
    else:
        MRT_Name.append(None)
        MRT_latitude.append(None)
        MRT_longitude.append(None)

In [78]:
DTL2_data = pd.DataFrame({'MRT_ID' : MRT_ID, 'MRT_station' : MRT_Name, 'MRT_latitude' : MRT_latitude, 'MRT_longitude' : MRT_longitude})
DTL2_data

,MRT_ID,MRT_station,MRT_latitude,MRT_longitude
0,DT1,BUKIT PANJANG MRT STATION (DT1),1.37900211641036,103.761535114929
1,DT2,CASHEW MRT STATION (DT2),1.36981544925552,103.76443921414
2,DT3,HILLVIEW MRT STATION (DT3),1.36234486803558,103.767418254007
3,DT4,HUME MRT STATION (DT4),1.3545953239503,103.769118851887
4,DT5,BEAUTY WORLD MRT STATION (DT5),1.34122317558549,103.775794285083
5,DT6,KING ALBERT PARK MRT STATION (DT6),1.33566161262214,103.783399113441
6,DT7,SIXTH AVENUE MRT STATION (DT7),1.33085764536226,103.796906838288
7,DT8,TAN KAH KEE MRT STATION (DT8),1.32555623519898,103.807760304729
8,DT9,BOTANIC GARDENS MRT STATION (DT9),1.32248807089544,103.816013994259
9,CC19,BOTANIC GARDENS MRT STATION (CC19),1.32211373558877,103.814984510627


To run the DiD regression model, I will only use HDB data from towns that the DTL Stage 2 line runs through, which are bukit panjang, choa chu kang, bukit batok, and bukit timah. 

Towns which are too far away from the DTL Stage 2 stations may have systematically different housing markets. This violates the parallel trends assumption, which is critical in a DiD model.

The control group should be as similar as possible to the treatment group, except for the treatment itself (proximity to DTL2 stations).

In [79]:
# Filter dataset to only use HDB resale flats in selected towns as model data
selected_towns = ['BUKIT PANJANG', 'CHOA CHU KANG', 'BUKIT BATOK', 'BUKIT TIMAH']
HDB_model_data = HDB_data_final[HDB_data_final['town'].isin(selected_towns)].copy()

To better isolate the effects of treatment, price_per_sqm will be used instead of resale price.

The natural log of price_per_sqm, ln(price_per_sqm), will be used instead of price_per_sqm as output variable to reduce heteroskedasticity and skewness. 

Heteroskedasticity means that the variance of errors is not constant across all observations which is common in HDB resale prices. For example, 5-Room or Executive flats may have a different variability in price_per_sqm than smaller 2-Room flats.

In [80]:
HDB_model_data['price_per_sqm'] = HDB_model_data['resale_price'] / HDB_model_data['floor_area_sqm']
HDB_model_data['ln_price'] = np.log(HDB_model_data['price_per_sqm'])

The DTL2 was officially opened on Dec 27, 2015 (rounded up to 2016) and will be used to create the post binary variable.

In [81]:
HDB_model_data['post'] = (HDB_model_data['transaction_year'] >= 2016).astype(int)  

## 3. Apply DiD Regression Model for Years 2011 to 2020

The dataset will be filtered to only use HDB resale flats transacted between 2011 and 2020 for my first DiD model.

In [82]:
HDB_model1_data = HDB_model_data[(HDB_model_data['transaction_year'] > 2010) & (HDB_model_data['transaction_year'] <= 2020)]
HDB_model1_data.shape

(24430, 39)

3 different proximities from DTL2 stations will be tested using the DiD regression model: 250m, 500m, and 750m.

In [83]:
# Set radius of search for HDB flats
proximity_list = [250, 500, 750]
proximity_dfs = {}

# Loop through proximity list
for proximity in proximity_list:
    temp_df = HDB_model1_data.copy()
    temp_df['treatment'] = 0

# Loop through HDB dataset
    for HDB_index, HDB_row in temp_df.iterrows():
        HDB_coords = (HDB_row['HDB_latitude'], HDB_row['HDB_longitude'])
        
        # Loop through MRT stations
        for MRT_index, MRT_row in DTL2_data.iterrows():
            MRT_coords = (MRT_row['MRT_latitude'], MRT_row['MRT_longitude'])
            distance = geodesic(HDB_coords, MRT_coords).meters
            
            # If within range of any station assign '1' to treatment column and stop looping through other MRT stations
            if distance <= proximity:
                temp_df.loc[HDB_index, 'treatment'] = 1
                break

    # Store the DataFrame in dictionary keyed by proximity
    proximity_dfs[proximity] = temp_df

HDB_model1_data_250m = proximity_dfs[250]
HDB_model1_data_500m = proximity_dfs[500]
HDB_model1_data_750m = proximity_dfs[750]

The HDB flats that were within a proximity of 250 m, 500 m, and 750 m from the DTL2 stations were assigned '1' as the treatment group, and stored in separate dataframes.

In [84]:
print(HDB_model1_data_250m['treatment'].value_counts())
print(HDB_model1_data_500m['treatment'].value_counts())
print(HDB_model1_data_750m['treatment'].value_counts())

treatment
0    24105
1      325
Name: count, dtype: int64
treatment
0    23096
1     1334
Name: count, dtype: int64
treatment
0    21339
1     3091
Name: count, dtype: int64


### 3.1. Pre-trend and parallel trend check

Based on results from regression models developed in Section 2, the following features are to be added as covariates for the DiD regression model: 'town', 'flat_model', and 'floor'.

Due to multi-collinearity considerations, variables 'flat_age', 'flat_type', 'floor_area_sqm', 'transaction_year', 'lease_commence_date', and 'transaction_year_month' will not be included.

Adding covariates is important to account for confounding factors so as to better isolate the effect of the treatment (proximity to DTL2 stations).

In [85]:
proximity_dfs = {250 : HDB_model1_data_250m, 500 : HDB_model1_data_500m, 750 : HDB_model1_data_750m}
pretrend_results = {}

for proximity, df in proximity_dfs.items():
        
    # Filter pre-treatment years
    pre_trends = df[df['transaction_year'] < 2016]
    
    # Define pre-trend regression formula and fit model
    pre_trend_formula = 'ln_price ~ treatment*C(transaction_year) + floor + C(town) + C(flat_model)'
    pre_trend_model = smf.ols(pre_trend_formula, data = pre_trends).fit(cov_type = 'HC3')
    
    # Extract treatment × year coefficients and p-values
    pre_coefs = pre_trend_model.params.filter(like = 'treatment:C(transaction_year)')
    pre_pvals = pre_trend_model.pvalues.filter(like = 'treatment:C(transaction_year)')
    
    df_results = pd.DataFrame({'term' : pre_coefs.index,
                               'coef' : pre_coefs.values,
                               '%_difference' : ((np.exp(pre_coefs.values) - 1) * 100).round(2),
                               'p_value' : pre_pvals.values})
    
    # Store the results in dictionary keyed by proximity
    pretrend_results[proximity] = df_results

pretrend_250m = pretrend_results[250]
pretrend_500m = pretrend_results[500]
pretrend_750m = pretrend_results[750]

cov_type = 'HC3' refers to the method for computation of standard errors. HC3 method is robust to heteroskedasticity.

In [86]:
print('====250 m====')
print(pretrend_250m)
print('\n====500 m====')
print(pretrend_500m)
print('\n====750 m====')
print(pretrend_750m)

====250 m====
                                    term      coef  %_difference   p_value
0  treatment:C(transaction_year)[T.2012]  0.024941          2.53  0.314733
1  treatment:C(transaction_year)[T.2013]  0.007341          0.74  0.804752
2  treatment:C(transaction_year)[T.2014]  0.060277          6.21  0.007861
3  treatment:C(transaction_year)[T.2015]  0.115247         12.22  0.000002

====500 m====
                                    term      coef  %_difference       p_value
0  treatment:C(transaction_year)[T.2012] -0.009101         -0.91  3.820303e-01
1  treatment:C(transaction_year)[T.2013] -0.015978         -1.59  1.790396e-01
2  treatment:C(transaction_year)[T.2014]  0.044172          4.52  9.987064e-05
3  treatment:C(transaction_year)[T.2015]  0.085376          8.91  1.147253e-10

====750 m====
                                    term      coef  %_difference   p_value
0  treatment:C(transaction_year)[T.2012] -0.006080         -0.61  0.429443
1  treatment:C(transaction_year)[T.2

For all 3 proximity models, pre-treatment differences exist: For 250 m, differences are significant (p-value < 0.05) for 2014 and 2015. For 500 m, differences are significant for 2014 and 2015. For 750 m, differences are significant for 2013 and 2015. 

This indicates that the parallel trends assumption may be violated slightly, hence treatment and control were not perfectly moving together in all pre-treatment years.

For 750 m, the coefficients are relatively small in magnitude (~4%), so the divergence may not be huge. For 250m and 500m, cofficients are relatively larger (~5-12%), so larger divergence will be expected.

In [87]:
# Create combined subplots from the 3 proximity dataframes
fig11 = make_subplots(rows = 1, cols = 3, subplot_titles = [f"{p} m" for p in proximity_dfs.keys()])

# Loop through proximities and plot treatment vs control
for i, (proximity, df) in enumerate(proximity_dfs.items(), start = 1):
    avg_prices = df.groupby(['transaction_year', 'treatment'])['price_per_sqm'].median().reset_index()
    
    # Control group
    control = avg_prices[avg_prices['treatment'] == 0]
    fig11.add_trace(go.Scatter(x = control['transaction_year'], y = control['price_per_sqm'], 
                              mode = 'lines+markers', name = 'Control', 
                              line = dict(color='blue'), showlegend = (i==1),
                              hovertemplate = 'Year: %{x}<br>'+'Median Price per SQM: S$%{y:.0f} per m²<br>'+'Group: Control<extra></extra>'),
                              row = 1, col = i)
   
    # Treatment group
    treatment = avg_prices[avg_prices['treatment'] == 1]
    fig11.add_trace(go.Scatter(x = treatment['transaction_year'], y = treatment['price_per_sqm'],
                              mode = 'lines+markers', name = 'Treatment',
                              line = dict(color='red'), showlegend = (i == 1),
                              hovertemplate = 'Year: %{x}<br>'+'Median Price per SQM: S$%{y:.0f} per m²<br>'+'Group: Treatment<extra></extra>'),
                              row = 1, col = i)
        
    # Line indicating date of DTL2 opening (Dec 27, 2015)
    fig11.add_trace(go.Scatter(x = [2015, 2015], y = [avg_prices['price_per_sqm'].min(), avg_prices['price_per_sqm'].max()],
                              mode = 'lines', line=dict(color='black', dash='dash'),
                              name = 'DTL2 Opening', showlegend=(i == 1)),
                              row = 1, col = i)
    
    fig11.update_xaxes(title_text = 'Transaction Year', row = 1, col = i)

fig11.update_layout(height = 500, width = 1600,
                   title_text = 'Parallel Trends: Treatment vs Control Group for Different Proximities (2011 to 2020)', title_x = 0.5,
                   yaxis_title = 'Median Price per SQM (S$/m²)')

fig11.show()

By visual inspection, the parallel trend assumption broadly holds for all 3 proximities.

A divergence in mean price per sqm between treatment and control group was observed for all 3 proximities after DTL2 opening date on Dec 27, 2015, signifiying an increase in mean price per sqm of HDB resale flats for treatment group relative to control group, after DTL2 opening.

### 3.2. Intepret Results of DiD regression model

In [88]:
all_results = []

# Loop through each proximity DataFrame
for proximity, df in proximity_dfs.items():
    
    # Define DiD formula and fit model
    DiD_formula = 'ln_price ~ post + treatment + post:treatment + floor + C(town) + C(flat_model)'
    DiD_model = smf.ols(DiD_formula, data = df).fit(cov_type = 'HC3')
    
    # Extract treatment effect and p-value
    beta_3 = DiD_model.params['post:treatment']
    pval = DiD_model.pvalues['post:treatment']
    effect_pct = (np.exp(beta_3) - 1) * 100
    
    # Append summary info
    all_results.append({'Start Year' : 2011,
                        'End Year' : 2020,
                        'Proximity (m)' : proximity,
                        'Price_per_sqm Increase (%)' : f'{effect_pct:.2f}',
                        'p_value' : pval})

summary_df = pd.DataFrame(all_results)
summary_df

,Start Year,End Year,Proximity (m),Price_per_sqm Increase (%),p_value
0,2011,2020,250,13.28,8.453755e-25
1,2011,2020,500,10.78,9.258313e-69
2,2011,2020,750,6.76,4.712811e-48


On further inspection of the p-value and % increase in price per sqm, I confirm that there is a highly significant (p < 0.01) statistical difference between resale prices of HDB flats served by the DTL2 MRT stations.

There was a 13.28 % increase in price per sqm of resale flats that are within a 250 m radius of a DTL2 station, 5 years after vs before its opening, relative to resale flats that are further away. <br>
There was a 10.78 % increase in price per sqm of resale flats that are within a 500 m radius of a DTL2 station, 5 years after vs before its opening, relative to resale flats that are further away.<br>
There was a 6.76 % increase in price per sqm of resale flats that are within a 750 m radius of a DTL2 station, 5 years after vs before its opening, relative to resale flats that are further away. <br>

The increasing price difference of resale prices highlights another point that flats in closer proximity to the DTL2 stations experienced a greater price increase. <Br>

## 4. Apply DiD Regression Model for Years 2013 to 2018

The dataset will be filtered to only use HDB resale flats transacted between 2013 and 2028 for my second DiD model.

In [89]:
HDB_model2_data = HDB_model_data[(HDB_model_data['transaction_year'] > 2012) & (HDB_model_data['transaction_year'] <= 2018)]
HDB_model2_data.shape

(13078, 39)

The same 3 proximities used in the first model will be tested using the 2nd DiD regression model: 250 m, 500 m, and 750 m.

In [90]:
# Set radius of search for HDB flats
proximity_list = [250, 500, 750]
proximity_dfs = {}

# Loop through proximity list
for proximity in proximity_list:
    temp_df = HDB_model2_data.copy()
    temp_df['treatment'] = 0

# Loop through HDB dataset
    for HDB_index, HDB_row in temp_df.iterrows():
        HDB_coords = (HDB_row['HDB_latitude'], HDB_row['HDB_longitude'])
        
        # Loop through MRT stations
        for MRT_index, MRT_row in DTL2_data.iterrows():
            MRT_coords = (MRT_row['MRT_latitude'], MRT_row['MRT_longitude'])
            distance = geodesic(HDB_coords, MRT_coords).meters
            
            # If within range of any station assign '1' to treatment column and stop looping through other MRT stations
            if distance <= proximity:
                temp_df.loc[HDB_index, 'treatment'] = 1
                break

    # Store the DataFrame in dictionary keyed by proximity
    proximity_dfs[proximity] = temp_df

HDB_model2_data_250m = proximity_dfs[250]
HDB_model2_data_500m = proximity_dfs[500]
HDB_model2_data_750m = proximity_dfs[750]

The HDB flats that were within a proximity of 250 m, 500 m, and 750 m from the DTL2 stations were assigned '1' as the treatment group, and stored in separate dataframes.

In [91]:
print(HDB_model2_data_250m['treatment'].value_counts())
print(HDB_model2_data_500m['treatment'].value_counts())
print(HDB_model2_data_750m['treatment'].value_counts())

treatment
0    12884
1      194
Name: count, dtype: int64
treatment
0    12326
1      752
Name: count, dtype: int64
treatment
0    11271
1     1807
Name: count, dtype: int64


### 4.1. Pre-trend and parallel trend check

In [92]:
proximity_dfs = {250 : HDB_model2_data_250m, 500 : HDB_model2_data_500m, 750 : HDB_model2_data_750m}
pretrend_results = {}

for proximity, df in proximity_dfs.items():
        
    # Filter pre-treatment years
    pre_trends = df[df['transaction_year'] < 2016]
    
    # Define pre-trend regression formula and fit model
    pre_trend_formula = 'ln_price ~ treatment*C(transaction_year) + floor + C(town) + C(flat_model)'
    pre_trend_model = smf.ols(pre_trend_formula, data = pre_trends).fit(cov_type = 'HC3')
    
    # Extract treatment × year coefficients and p-values
    pre_coefs = pre_trend_model.params.filter(like='treatment:C(transaction_year)')
    pre_pvals = pre_trend_model.pvalues.filter(like='treatment:C(transaction_year)')
    
    df_results = pd.DataFrame({'term' : pre_coefs.index,
                               'coef' : pre_coefs.values,
                               '%_difference' : ((np.exp(pre_coefs.values) - 1) * 100).round(2),
                               'p_value' : pre_pvals.values})
    
    # Store the results in dictionary keyed by proximity
    pretrend_results[proximity] = df_results

pretrend_250m = pretrend_results[250]
pretrend_500m = pretrend_results[500]
pretrend_750m = pretrend_results[750]

In [93]:
print('====250 m====')
print(pretrend_250m)
print('\n====500 m====')
print(pretrend_500m)
print('\n====750 m====')
print(pretrend_750m)

====250 m====
                                    term      coef  %_difference   p_value
0  treatment:C(transaction_year)[T.2014]  0.049971          5.12  0.080372
1  treatment:C(transaction_year)[T.2015]  0.106392         11.23  0.000295

====500 m====
                                    term      coef  %_difference       p_value
0  treatment:C(transaction_year)[T.2014]  0.064231          6.63  2.180711e-07
1  treatment:C(transaction_year)[T.2015]  0.101586         10.69  7.724893e-14

====750 m====
                                    term      coef  %_difference       p_value
0  treatment:C(transaction_year)[T.2014]  0.036844          3.75  6.525480e-05
1  treatment:C(transaction_year)[T.2015]  0.060233          6.21  1.444318e-10


For all 3 proximity models, pre-treatment differences exist: Differences are significant (p-value < 0.05) for almost all years and proximities except for year 2014 for 250 m proximity.

This indicates that the parallel trends assumption may be violated slightly, hence treatment and control were not perfectly moving together in all pre-treatment years.

For 750 m, the coefficients are relatively small in magnitude (~3-6%), so the divergence may not be huge. For 250 m and 500 m, cofficients are relatively larger (~5-11%), so larger divergence will be expected.

In [94]:
# Create combined subplots from the 3 proximity dataframes
fig12 = make_subplots(rows = 1, cols = 3, subplot_titles = [f"{p} m" for p in proximity_dfs.keys()])

# Loop through proximities and plot treatment vs control
for i, (proximity, df) in enumerate(proximity_dfs.items(), start = 1):
    avg_prices = df.groupby(['transaction_year', 'treatment'])['price_per_sqm'].median().reset_index()
    
    # Control group
    control = avg_prices[avg_prices['treatment'] == 0]
    fig12.add_trace(go.Scatter(x = control['transaction_year'], y = control['price_per_sqm'], 
                              mode = 'lines+markers', name = 'Control', 
                              line = dict(color='blue'), showlegend = (i==1),
                              hovertemplate = 'Year: %{x}<br>'+'Median Price per SQM: S$%{y:.0f} per m²<br>'+'Group: Control<extra></extra>'), 
                              row = 1, col = i)
    
    # Treatment group
    treatment = avg_prices[avg_prices['treatment'] == 1]
    fig12.add_trace(go.Scatter(x = treatment['transaction_year'], y = treatment['price_per_sqm'],
                             mode = 'lines+markers', name = 'Treatment',
                             line = dict(color='red'), showlegend = (i == 1),
                             hovertemplate = 'Year: %{x}<br>'+'Median Price per SQM: S$%{y:.0f} per m²<br>'+'Group: Treatment<extra></extra>'), 
                             row = 1, col = i)
    
    # Line indicating date of DTL2 opening (Dec 27, 2015)
    fig12.add_trace(go.Scatter(x = [2015, 2015], y = [avg_prices['price_per_sqm'].min(), avg_prices['price_per_sqm'].max()],
                               mode = 'lines', line = dict(color = 'black', dash = 'dash'),
                               name = 'DTL2 Opening', showlegend = (i == 1)),
                               row = 1, col = i)
    
    fig12.update_xaxes(title_text = 'Transaction Year', row = 1, col = i)

fig12.update_layout(height = 500, width = 1600,
                    title_text = 'Parallel Trends: Treatment vs Control Group for Different Proximities (2013 to 2018)', title_x = 0.5,
                    yaxis_title = 'Median Price per SQM (S$/m²)')

fig12.show()

By visual inspection, the parallel trend assumption broadly holds for all 3 proximities.

A divergence in mean price per sqm between treatment and control group was observed for all 3 proximities after DTL2 opening date on Dec 27, 2015, signifiying an increase in mean price per sqm of HDB resale flats for treatment group relative to control group, after DTL2 opening.

### 4.2. Intepret Results of DiD regression model

In [95]:
all_results = []

# Loop through each proximity DataFrame
for proximity, df in proximity_dfs.items():
    
    # Define DiD formula and fit model
    DiD_formula = 'ln_price ~ post + treatment + post:treatment + floor + C(town) + C(flat_model)'
    DiD_model = smf.ols(DiD_formula, data = df).fit(cov_type = 'HC3')
    
    # Extract treatment effect and p-value
    beta_3 = DiD_model.params['post:treatment']
    pval = DiD_model.pvalues['post:treatment']
    effect_pct = (np.exp(beta_3) - 1) * 100
    
    # Append summary info
    all_results.append({'Start Year' : 2013,
                        'End Year' : 2018,
                        'Proximity (m)' : proximity,
                        'Price_per_sqm Increase (%)' : f'{effect_pct:.2f}',
                        'p_value' : pval})

summary_df = pd.DataFrame(all_results)
summary_df

,Start Year,End Year,Proximity (m),Price_per_sqm Increase (%),p_value
0,2013,2018,250,10.56,1.250310e-11
1,2013,2018,500,8.40,4.211350e-26
2,2013,2018,750,5.51,3.130666e-20


On further inspection of the p-value and % increase in price per sqm, I confirm that there is a highly significant (p < 0.01) statistical difference between resale prices of HDB flats served by the DTL2 MRT stations.

There was a 10.56 % increase in price per sqm of resale flats that are within a 250 m radius of a DTL2 station, 3 years after vs before its opening, relative to resale flats that are further away. <br>
There was a 8.40 % increase in price per sqm of resale flats that are within a 500 m radius of a DTL2 station, 3 years after vs before its opening, relative to resale flats that are further away.<br>
There was a 5.51 % increase in price per sqm of resale flats that are within a 750 m radius of a DTL2 station, 3 years after vs before its opening, relative to resale flats that are further away. <br>

The increasing price difference of resale prices highlights another point that flats in closer proximity to the DTL2 stations experienced a greater price increase. <Br>

## 5. Conclusion

For this question, 2 DiD regression models for different observation periods of 2011 to 2020 and 2013 to 2018, each model testing 3 different proximities of 250, 500, and 750 m to DTL2 statinons. 

The model accounts for covariates such as town, flat model, floor area, and floor level of resale flats.

The result demonstrates very clearly that a highly significant increase in resale price_per_sqm of HDB resale flat was observed for flats that are in close proximity of a DTL2 station, after vs before its opening, relative to resale flats that are farther away.

Furthermore, the nearer the flats are to the DTL2 stations, the greater the increase in resale price per sqm of HDB flats served by these stations. <br><br>

# Question D***

There have been comments online that people are buying flats in towns further from the city so that the cost savings can be used for a car. Are resale prices in HDB estates in areas further away from the city (i.e. Sengkang and Punggol) impacted by Certificate of Entitlement (COE) prices for cars? <br>

## 1. Reframing the Question for Data Analysis

The data analysis should address the following points:

<b><i>How to determine if resale prices are impacted by COE prices for cars? </b></i><br>
Evaluate Car COE price coefficients of regression models correlating median resale prices per sqm to median COE prices <br>
Statistically significant Car COE effect on resale prices per sqm <br>

<b><i>What is the unit of comparison? </b></i><br>
Compare flats in suburban towns Sengkang and Punggol to flats in city towns Central Area and Kallang/Whampoa <br>

<b><i>What is the time frame? </b></i><br>
Analyze data from 2010 to 2025, limited by the time range of COE data from data.gov.sg <br>

Data Analysis Question: <b><i> Are the median resale prices per sqm of flats in Sengkang or Punggol more senitive to changes in Car COE prices as compared to flats in Central Area or Kallang/Whampoa?

## 2. Cleaning and Merging of COE Data

In [96]:
COE_data = pd.read_csv('./Datasets/COEBiddingResultsPrices.csv')

COE price data was downloaded from https://data.gov.sg/datasets?query=COE&resultId=d_69b3380ad7e51aff3a7dcc84eba52b8a&page=1&dataExplorerPage=1

In [97]:
COE_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1850 entries, 0 to 1849
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   month          1850 non-null   object
 1   bidding_no     1850 non-null   int64 
 2   vehicle_class  1850 non-null   object
 3   quota          1850 non-null   int64 
 4   bids_success   1850 non-null   object
 5   bids_received  1850 non-null   object
 6   premium        1850 non-null   int64 
dtypes: int64(3), object(4)
memory usage: 101.3+ KB


In [98]:
COE_data

,month,bidding_no,vehicle_class,quota,bids_success,bids_received,premium
0,2010-01,1,Category A,1152,1145,1342,18502
1,2010-01,1,Category B,687,679,883,19190
2,2010-01,1,Category C,173,173,265,19001
3,2010-01,1,Category D,373,365,509,889
4,2010-01,1,Category E,586,567,1011,19889
...,...,...,...,...,...,...,...
1845,2025-08,2,Category A,1264,"1,257","1,922",104524
1846,2025-08,2,Category B,792,789,"1,221",124400
1847,2025-08,2,Category C,273,271,472,72190
1848,2025-08,2,Category D,540,535,654,8809


The dataset contains 1850 rows and 7 columns.

The COE price information is contained in the 'premium' column.

COE data from data.gov.sg only starts from 2010. Comparison between HDB resale prices and COE prices can only be done from 2010 to 2025.

In [99]:
COE_data = COE_data[(COE_data['vehicle_class'] == 'Category A') | (COE_data['vehicle_class'] == 'Category B')]
COE_data

,month,bidding_no,vehicle_class,quota,bids_success,bids_received,premium
0,2010-01,1,Category A,1152,1145,1342,18502
1,2010-01,1,Category B,687,679,883,19190
5,2010-01,2,Category A,1151,1149,1673,20501
6,2010-01,2,Category B,717,717,1105,22400
10,2010-02,1,Category A,1154,1153,1326,19989
...,...,...,...,...,...,...,...
1836,2025-07,2,Category B,784,782,"1,148",119101
1840,2025-08,1,Category A,1268,"1,257","1,755",102009
1841,2025-08,1,Category B,789,789,"1,271",123498
1845,2025-08,2,Category A,1264,"1,257","1,922",104524


Dataset was filtered to only keep Categories A and B which are strictly for cars. The other categories are for other vehicle types.

Cat A: Cars ≤1600cc & ≤97kW (non-electric) / electric cars ≤110kW <br>
Cat B: Cars >1600cc or >97kW (non-electric) / electric cars >110kW

In [100]:
Car_COE_data = (COE_data.groupby(['month', 'vehicle_class'])['premium'].mean().reset_index())
Car_COE_data

,month,vehicle_class,premium
0,2010-01,Category A,19501.5
1,2010-01,Category B,20795.0
2,2010-02,Category A,20164.5
3,2010-02,Category B,23534.5
4,2010-03,Category A,24595.5
...,...,...,...
365,2025-06,Category B,114835.0
366,2025-07,Category A,101102.0
367,2025-07,Category B,119350.5
368,2025-08,Category A,103266.5


There are 2 COE biddings per month. Since our HDB data is only granular to the month level, the COE data had to be converted to the same level of granularity by averaging the 2 bidding prices for both Cat A and B.

In [101]:
Car_COE_data = Car_COE_data.pivot(index = 'month', columns = 'vehicle_class', values = 'premium').reset_index()
Car_COE_data = Car_COE_data.rename(columns = {'Category A': 'CatA_premium', 'Category B': 'CatB_premium'})
Car_COE_data.columns.name = None
Car_COE_data

,month,CatA_premium,CatB_premium
0,2010-01,19501.5,20795.0
1,2010-02,20164.5,23534.5
2,2010-03,24595.5,31239.0
3,2010-04,32000.5,42751.0
4,2010-05,26745.5,36300.0
...,...,...,...
180,2025-04,98612.0,117451.0
181,2025-05,102755.0,118439.0
182,2025-06,97561.5,114835.0
183,2025-07,101102.0,119350.5


Car COE prices for both Cat A and B are relevant for my analysis. For easier merging with the HDB dataset, the Car_COE_date was pivoted so that both Cat A and B premiums are now columns.

In [102]:
HDB_data = HDB_data_final.copy()
HDB_data.loc[:, 'transaction_year_month'] = pd.to_datetime(HDB_data['transaction_year_month'])
Car_COE_data.loc[:, 'month'] = pd.to_datetime(Car_COE_data['month'])

Converted both HDB_data 'transaction_year_month' and Car_COE_data 'month' columns into similar format to be used as joining key for merging the 2 dataset.

In [103]:
HDB_COE_data = pd.merge(HDB_data, Car_COE_data, left_on = 'transaction_year_month', right_on = 'month', how = 'inner')
HDB_COE_data = HDB_COE_data.drop(columns = ['month'])
HDB_COE_data

,transaction_year_month,transaction_year,lease_commence_date,flat_age,flat_type,flat_model,floor_area_sqm,storey_range,floor,resale_price,...,prisch_longitude,Distance_m_Nearest_prisch,Nearest_secsch,secsch_latitude,secsch_longitude,Distance_m_Nearest_secsch,price_per_sqm,region,CatA_premium,CatB_premium
0,2010-01-01 00:00:00,2010,1977,33,2 ROOM,IMPROVED,44.0,10 TO 12,11,202000.0,...,103.851010,443.709276,DEYI SECONDARY SCHOOL,1.365997,103.852546,577.756300,4590.909091,Central,19501.5,20795.0
1,2010-01-01 00:00:00,2010,1978,32,2 ROOM,IMPROVED,44.0,01 TO 03,2,208000.0,...,103.851010,120.999221,DEYI SECONDARY SCHOOL,1.365997,103.852546,274.933707,4727.272727,Central,19501.5,20795.0
2,2010-01-01 00:00:00,2010,1978,32,2 ROOM,IMPROVED,44.0,07 TO 09,8,180000.0,...,103.851010,120.999221,DEYI SECONDARY SCHOOL,1.365997,103.852546,274.933707,4090.909091,Central,19501.5,20795.0
3,2010-01-01 00:00:00,2010,1978,32,2 ROOM,IMPROVED,44.0,07 TO 09,8,180000.0,...,103.851010,120.999221,DEYI SECONDARY SCHOOL,1.365997,103.852546,274.933707,4090.909091,Central,19501.5,20795.0
4,2010-01-01 00:00:00,2010,1977,33,2 ROOM,IMPROVED,44.0,01 TO 03,2,198000.0,...,103.851010,443.709276,DEYI SECONDARY SCHOOL,1.365997,103.852546,577.756300,4500.000000,Central,19501.5,20795.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
360494,2025-04-01 00:00:00,2025,1988,37,EXECUTIVE,MAISONETTE,146.0,04 TO 06,5,975000.0,...,103.830143,393.428838,NAVAL BASE SECONDARY SCHOOL,1.419379,103.831050,506.837987,6678.082192,North,98612.0,117451.0
360495,2025-05-01 00:00:00,2025,1988,37,EXECUTIVE,APARTMENT,142.0,04 TO 06,5,1000000.0,...,103.830143,393.428838,NAVAL BASE SECONDARY SCHOOL,1.419379,103.831050,506.837987,7042.253521,North,102755.0,118439.0
360496,2025-07-01 00:00:00,2025,1987,38,EXECUTIVE,MAISONETTE,146.0,04 TO 06,5,980000.0,...,103.830143,537.060102,ORCHID PARK SECONDARY SCHOOL,1.415320,103.837859,536.099259,6712.328767,North,101102.0,119350.5
360497,2025-05-01 00:00:00,2025,1987,38,MULTI GENERATION,MULTI GENERATION,147.0,04 TO 06,5,945000.0,...,103.839101,294.884175,CHUNG CHENG HIGH SCHOOL (YISHUN),1.419226,103.837391,212.737676,6428.571429,North,102755.0,118439.0


The 2 datasets were merged using 'inner', deleting away all data before 2010-01 because that is the earliest date available for the Car_COE_data.<br>

## 3. Visualisation of COE and HDB Prices over Time

Median COE prices for both CatA and CatB will be plotted together with HDB median resale prices per sqm from 2010 to 2025 to better visualise the relationship between these variables. The prices are aggregated by median on a monthly basis.

In [104]:
HDB_COE_data['price_per_sqm'] = HDB_COE_data['resale_price'] / HDB_COE_data['floor_area_sqm']

Resale price per sqm will be used as a metric over resale price to better account for flat size differences and reduce price variance for larger flats.

In [105]:
HDB_COE_Sengkang = HDB_COE_data[HDB_COE_data['town'] == 'SENGKANG']
median_price_month = HDB_COE_Sengkang.groupby('transaction_year_month').agg({'price_per_sqm' : 'median', 'CatB_premium' : 'median', 'CatA_premium' : 'median'}).reset_index()

The data will be separately visualised for 4 HDB towns. 

Sengkang and Punggol which are suburban towns far away from the city and Central Area and Kallang/Whampoa which are in the city center.

In [106]:
fig13 = go.Figure()

# Line plot of HDB median resale price per sqm
fig13.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['price_per_sqm'],
                           name = 'Median Price per SQM', mode = 'lines+markers',
                           hovertemplate = 'Date: %{x}<br>Median Price per SQM: S$%{y:,.0f} per m²<br><extra></extra>',
                           line = dict(color = 'blue')))

# Line plot of Median COE Price (Category A) on secondary y-axis
fig13.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['CatA_premium'],
                           name = 'Median COE Price (Cat A)', mode = 'lines+markers',
                           line = dict(color = 'red'),
                           hovertemplate = 'Date: %{x}<br>Median COE Price (Cat A): S$%{y:,.0f}<br><extra></extra>',
                           yaxis = 'y2'))

# Line plot of Median COE Price (Category B) on secondary y-axis
fig13.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['CatB_premium'],
                           name = 'Median COE Price (Cat B)',
                           mode = 'lines+markers',
                           line = dict(color = 'green'),
                           hovertemplate = 'Date: %{x}<br>Median COE Price (Cat B): S$%{y:,.0f}<br><extra></extra>',
                           yaxis = 'y2'))

# Update layout for secondary y-axis
fig13.update_layout(title = {'text' : 'SengKang: HDB Median Resale Price per SQM vs <br>Median COE Price (CatA and B) over Time', 'x' : 0.5, 'y' : 0.95},      
                    xaxis_title = 'Date',
                    yaxis = dict(title = 'Median Price per SQM (S$/m²)'),
                    yaxis2 = dict(title = 'Median COE Price (CatA and B) (S$)', overlaying = 'y', side = 'right'),
                    width = 800, height = 500,
                    legend = dict(x = 0.02, y = 0.95),
                    margin = dict(l = 80, r = 80, t = 65, b = 65))

fig13.show()

From visual inspection of the Sengkang plots, all 3 prices appear to follow a time trend and run in parallel to each other, rising and falling together.

This might signify that all three prices are being influenced by broader macro factors such as inflation, policy changes, or demand/supply pressures. For example, during economic growth, both COE prices and HDB resale prices tend to rise due to increase in demand for both cars and housing.

This possibly indicates that after accounting for influences from the time trend, regression coefficients of COE prices might become small or insignificant.

In [107]:
HDB_COE_Punggol = HDB_COE_data[HDB_COE_data['town'] == 'PUNGGOL']
median_price_month = HDB_COE_Punggol.groupby('transaction_year_month').agg({'price_per_sqm' : 'median', 'CatB_premium' : 'median', 'CatA_premium' : 'median'}).reset_index()

In [108]:
fig14 = go.Figure()

# Line plot of HDB median resale price per sqm
fig14.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['price_per_sqm'],
                           name = 'Median Price per SQM', mode = 'lines+markers',
                           hovertemplate = 'Date: %{x}<br>Median Price per SQM: S$%{y:,.0f} per m²<br><extra></extra>',
                           line = dict(color = 'blue')))

# Line plot of Median COE Price (Category A) on secondary y-axis
fig14.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['CatA_premium'],
                           name = 'Median COE Price (Cat A)', mode = 'lines+markers',
                           line = dict(color = 'red'),
                           hovertemplate = 'Date: %{x}<br>Median COE Price (Cat A): S$%{y:,.0f}<br><extra></extra>',
                           yaxis = 'y2'))

# Line plot of Median COE Price (Category B) on secondary y-axis
fig14.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['CatB_premium'],
                           name = 'Median COE Price (Cat B)',
                           mode = 'lines+markers',
                           line = dict(color = 'green'),
                           hovertemplate = 'Date: %{x}<br>Median COE Price (Cat B): S$%{y:,.0f}<br><extra></extra>',
                           yaxis = 'y2'))

# Update layout for secondary y-axis
fig14.update_layout(title = {'text' : 'Punggol: HDB Median Resale Price per SQM vs <br>Median COE Price (CatA and B) over Time', 'x' : 0.5, 'y' : 0.95},      
                    xaxis_title = 'Date',
                    yaxis = dict(title = 'Median Price per SQM (S$/m²)'),
                    yaxis2 = dict(title = 'Median COE Price (CatA and B) (S$)', overlaying = 'y', side = 'right'),
                    width = 800, height = 500,
                    legend = dict(x = 0.02, y = 0.95),
                    margin = dict(l = 80, r = 80, t = 65, b = 65))

fig14.show()

For Punggol, the changes in median price per sqm deviates more from the COE prices as compared to Sengkang. Broadly, however, all 3 prices appear to follow a similar time trend.

This possibly indicates that after accounting for influences from the time trend, regression coefficients of COE prices might become small or insignificant.

In [109]:
HDB_COE_Kallang = HDB_COE_data[HDB_COE_data['town'] == 'KALLANG/WHAMPOA']
median_price_month = HDB_COE_Kallang.groupby('transaction_year_month').agg({'price_per_sqm' : 'median', 'CatB_premium' : 'median', 'CatA_premium' : 'median'}).reset_index()

In [110]:
fig15 = go.Figure()

# Line plot of HDB median resale price per sqm
fig15.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['price_per_sqm'],
                           name = 'Median Price per SQM', mode = 'lines+markers',
                           hovertemplate = 'Date: %{x}<br>Median Price per SQM: S$%{y:,.0f} per m²<br><extra></extra>',
                           line = dict(color = 'blue')))

# Line plot of Median COE Price (Category A) on secondary y-axis
fig15.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['CatA_premium'],
                           name = 'Median COE Price (Cat A)', mode = 'lines+markers',
                           line = dict(color = 'red'),
                           hovertemplate = 'Date: %{x}<br>Median COE Price (Cat A): S$%{y:,.0f}<br><extra></extra>',
                           yaxis = 'y2'))

# Line plot of Median COE Price (Category B) on secondary y-axis
fig15.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['CatB_premium'],
                           name = 'Median COE Price (Cat B)',
                           mode = 'lines+markers',
                           line = dict(color = 'green'),
                           hovertemplate = 'Date: %{x}<br>Median COE Price (Cat B): S$%{y:,.0f}<br><extra></extra>',
                           yaxis = 'y2'))

# Update layout for secondary y-axis
fig15.update_layout(title = {'text' : 'Kallang/Whampoa: HDB Median Resale Price per SQM vs <br>Median COE Price (CatA and B) over Time', 'x' : 0.5, 'y' : 0.95},      
                    xaxis_title = 'Date',
                    yaxis = dict(title = 'Median Price per SQM (S$/m²)'),
                    yaxis2 = dict(title = 'Median COE Price (CatA and B) (S$)', overlaying = 'y', side = 'right'),
                    width = 800, height = 500,
                    legend = dict(x = 0.02, y = 0.95),
                    margin = dict(l = 80, r = 80, t = 65, b = 65))

fig15.show()

Similar to Sengkang, all 3 prices appear to follow a highly similar time trend and run in parallel to each other, rising and falling together.

However, more month-to-month price fluctuations were observed likely because Kallang/Whampoa is a much older town and has fewer resale transactions per month. Median prices fluctuate depending on the mix of flats with different characteristics sold. Local market dynamics have a greater effect on prices.

In [111]:
HDB_COE_Central = HDB_COE_data[HDB_COE_data['town'] == 'CENTRAL AREA']
median_price_month = HDB_COE_Central.groupby('transaction_year_month').agg({'price_per_sqm' : 'median', 'CatB_premium' : 'median', 'CatA_premium' : 'median'}).reset_index()

In [112]:
fig16 = go.Figure()

# Line plot of HDB median resale price per sqm
fig16.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['price_per_sqm'],
                           name = 'Median Price per SQM', mode = 'lines+markers',
                           hovertemplate = 'Date: %{x}<br>Median Price per SQM: S$%{y:,.0f} per m²<br><extra></extra>',
                           line = dict(color = 'blue')))

# Line plot of Median COE Price (Category A) on secondary y-axis
fig16.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['CatA_premium'],
                           name = 'Median COE Price (Cat A)', mode = 'lines+markers',
                           line = dict(color = 'red'),
                           hovertemplate = 'Date: %{x}<br>Median COE Price (Cat A): S$%{y:,.0f}<br><extra></extra>',
                           yaxis = 'y2'))

# Line plot of Median COE Price (Category B) on secondary y-axis
fig16.add_trace(go.Scatter(x = median_price_month['transaction_year_month'],
                           y = median_price_month['CatB_premium'],
                           name = 'Median COE Price (Cat B)',
                           mode = 'lines+markers',
                           line = dict(color = 'green'),
                           hovertemplate = 'Date: %{x}<br>Median COE Price (Cat B): S$%{y:,.0f}<br><extra></extra>',
                           yaxis = 'y2'))

# Update layout for secondary y-axis
fig16.update_layout(title = {'text' : 'Central Area: HDB Median Resale Price per SQM vs <br>Median COE Price (CatA and B) over Time', 'x' : 0.5, 'y' : 0.95},      
                    xaxis_title = 'Date',
                    yaxis = dict(title = 'Median Price per SQM (S$/m²)'),
                    yaxis2 = dict(title = 'Median COE Price (CatA and B) (S$)', overlaying = 'y', side = 'right'),
                    width = 800, height = 500,
                    legend = dict(x = 0.02, y = 0.95),
                    margin = dict(l = 80, r = 80, t = 65, b = 65))

fig16.show()

In central area plots, median HDB prices per sqm experienced major month-to-month fluctuations, especially so from 2015 onwards. 

This is possibly due to smaller transaction volumes in central estates with fewer flats being sold monthly, so the median price can jump sharply if an expensive flat sells. The composition of units sold (size, age, floor level) can vary a lot month-to-month.

Difficult to identify any relationship between COE prices and resale prices from a visual plot.

## 4. Time Series Regression Analysis

A time series regression will be carried out separately for 4 HDB towns. 

Sengkang and Punggol which are suburban towns far away from the city and Central Area and Kallang/Whampoa which are in the city center.

In [113]:
towns = ['SENGKANG', 'PUNGGOL', 'CENTRAL AREA', 'KALLANG/WHAMPOA']

### 3.1. Effect of Cat A COE prices on HDB resale prices

Based on results from regression models developed in Section 2, the following features are to be added as covariates for a time-series regression model: 'flat_model', 'and 'floor'.

Due to multi-collinearity considerations, variables 'flat_age', 'flat_type', 'transaction_year', 'floor_area_sqm', and 'lease_commence_date' will not be included. 

'resale_price' will be the output variable and aggregated as a median value over a monthly time series.

Due to high correlation between 'CatA_premium' and 'CatB_premium' variables, they will be analyzed in different regression models. Also, running one model per COE category makes the marginal effect of that COE category on resale price more interpretable.

In [114]:
results = []

for town in towns:
    warnings.filterwarnings("ignore")
    town_df = HDB_COE_data[HDB_COE_data['town'] == town].copy()

    # One-hot encode categorical column flat_model
    flat_dummies = pd.get_dummies(town_df['flat_model'], prefix = 'flat_model')
    df = pd.concat([town_df[['transaction_year_month', 'price_per_sqm', 'CatA_premium', 'floor']], flat_dummies], axis = 1)

    # Aggregate by month using median for all values
    agg_model = df.groupby('transaction_year_month').agg('median').sort_index()
    agg_model = agg_model.asfreq('MS')
    agg_model = agg_model.dropna(subset = ['price_per_sqm'])
        
    # Define output (endog) and input (exog) variables
    endog = agg_model['price_per_sqm']
    exog = agg_model.drop(columns = ['price_per_sqm'])

    # Use auto_arima to find the best order (p, d, q) and seasonal order (P, D, Q, m)
    auto_model = pm.auto_arima(endog, exogenous = exog,
                               start_p = 0, start_q = 0,
                               max_p = 3, max_q = 3, d = None,                
                               seasonal = True, m = 12,                    
                               start_P = 0, start_Q = 0,
                               max_P = 2, max_Q = 2, D = None,                 
                               stepwise = True, information_criterion = 'aic',
                               suppress_warnings = True, error_action = 'ignore')

    # Fit SARIMAX with the best order and seasonal order
    final_model = SARIMAX(endog, exog = exog,
                          order = auto_model.order,
                          seasonal_order = auto_model.seasonal_order,
                          enforce_stationarity = False,
                          enforce_invertibility = False).fit()
    
    # Save results to list for conversion into dataframe
    results.append({'Town' : town,
                    'CatA_coef' : final_model.params.get('CatA_premium', None),
                    'CatA_p-value' : final_model.pvalues.get('CatA_premium', None)})

warnings.filterwarnings('default')

The SARIMAX model is perfect for this analysis because it has the capacity to include COE price as an exogenous regressor, allowing me to study the effect of COE on resale prices while controlling for other covariates.

The SARIMAX model also allows for autocorrelation and seasonal modeling which is important because both HDB rand COE prices tend to trend with time, if prices are high this month, they’re likely high next month.

An auto-arima function is used to select the best order and seasonal parameters based on the chosen information criterion 'AIC'. AIC stands for Akaike Information Criterion. It’s a metric used to compare statistical models.

In [115]:
results_df = pd.DataFrame(results)
results_df

,Town,CatA_coef,CatA_p-value
0,SENGKANG,0.000374,0.821671
1,PUNGGOL,0.000092,0.949144
2,CENTRAL AREA,0.016775,0.030157
3,KALLANG/WHAMPOA,-0.000073,0.988547


For Punggol, Sengkang, and Kallang/Whampoa resale flats, CatA COE coefficients are small and not statistically significant.

For Central Area resale flats, a $1 increase in median CatA COE premium is associated with a $0.017 increase in median price per sqm. As p-value < 0.05, this result is statistically significant. Despite, having the highest coefficient value of all 4 towns being compared, it is still extremely small.

### 4.2. Effect of Cat B COE prices on HDB resale prices

Due to high correlation between 'CatA_premium' and 'CatB_premium' variables, they were analyzed in different regression models. Also, running one model per COE category makes the marginal effect of that COE category on resale price more interpretable.

In [116]:
results = []

for town in towns:
    warnings.filterwarnings("ignore")
    town_df = HDB_COE_data[HDB_COE_data['town'] == town].copy()

    # One-hot encode categorical column flat_model
    flat_dummies = pd.get_dummies(town_df['flat_model'], prefix='flat_model')
    df = pd.concat([town_df[['transaction_year_month', 'price_per_sqm', 'CatB_premium', 'floor']], flat_dummies], axis = 1)

    # Aggregate by month using median for all values
    agg_model = df.groupby('transaction_year_month').agg('median').sort_index()
    agg_model = agg_model.asfreq('MS')
    agg_model = agg_model.dropna(subset = ['price_per_sqm'])
        
    # Define output (endog) and input (exog) variables
    endog = agg_model['price_per_sqm']
    exog = agg_model.drop(columns = ['price_per_sqm'])

    # Use auto_arima to find the best order (p, d, q) and seasonal order (P, D, Q, m)
    auto_model = pm.auto_arima(endog, exogenous = exog,
                               start_p = 0, start_q = 0,
                               max_p = 3, max_q = 3, d = None,                
                               seasonal = True, m = 12,                    
                               start_P = 0, start_Q = 0,
                               max_P = 2, max_Q = 2, D = None,                 
                               stepwise = True, information_criterion = 'aic',
                               suppress_warnings = True, error_action = 'ignore')

    # Fit SARIMAX with the best order and seasonal order
    final_model = SARIMAX(endog, exog = exog,
                          order = auto_model.order,
                          seasonal_order = auto_model.seasonal_order,
                          enforce_stationarity = False,
                          enforce_invertibility = False).fit()
    
    # Save results to list for conversion into dataframe
    results.append({'Town' : town,
                    'CatB_coef' : final_model.params.get('CatB_premium', None),
                    'CatB_p-value' : final_model.pvalues.get('CatB_premium', None)})

warnings.filterwarnings('default')

In [117]:
results_df = pd.DataFrame(results)
results_df

,Town,CatB_coef,CatB_p-value
0,SENGKANG,0.000328,0.786360
1,PUNGGOL,-0.000320,0.825938
2,CENTRAL AREA,0.013665,0.014802
3,KALLANG/WHAMPOA,-0.000683,0.859486


For Sengkang, Punggol and Kallang/Whampoa resale flats, CatB COE coefficients are small and not statistically significant.

However, for Central Area resale flats, a $1 increase in median CatA COE premium is associated with a $0.014 increase in median price per sqm. As p-value < 0.05, this result is statistically significant. Despite, having the highest coefficient value of all 4 towns being compared, it is still extremely small.

## 5. Conclusion

Based on results from the time regression analysis, after accounting for time trends and effects from covariates such as floor area, flat model, and floor level, the median price per sqm of resale flats in towns further away from the City, namely SengKang and Punggol, are not impacted by COE prices.

The same was observed for Kallang/Whampoa.

A statistically significant but marginal effect of COE prices on median price per sqm of resale flats in the central area was observed.

Overall, the impact of COE prices on HDB resale prices are not significant or marginal at best. <br><br>